### DISCUSSION TITLE
Cloudinary ImageUpload Error

### DISCUSSION THREAD

**Discussion by SelcoD** [neuefische/questions#649] [Web] [answered]
[PR](https://github.com/SelcoD/CampingFriends/pull/15)

I add the Cloudinary ImageUploader to my Project.
If you select a file and click Submit, it throws an Error.

It also happens when you go to formpage (by clicking "Add a new trip") and then refresh the page.

!!! with vercel it dosent show this error , in vercel preview nothing happens !!!

<img width="697" alt="Bildschirmfoto 2023-07-24 um 14 38 18" src="https://github.com/neuefische/questions/assets/132760864/886cc2a1-df89-407b-bdcd-fc49681fcb00">

The Error ->

Unhandled Runtime Error
Error: Hydration failed because the initial UI does not match what was rendered on the server.

Warning: Expected server HTML to contain a matching <form> in <div>.

See more info here: https://nextjs.org/docs/messages/react-hydration-error

Component Stack
form
div
ImageUpload
form
div
O
div
O
FormCard
FormPage
App
Call Stack
throwOnHydrationMismatch
node_modules/react-dom/cjs/react-dom.development.js (12507:0)
tryToClaimNextHydratableInstance
node_modules/react-dom/cjs/react-dom.development.js (12535:0)
updateHostComponent
node_modules/react-dom/cjs/react-dom.development.js (19902:0)
beginWork
node_modules/react-dom/cjs/react-dom.development.js (21618:0)
HTMLUnknownElement.callCallback
node_modules/react-dom/cjs/react-dom.development.js (4164:0)
Object.invokeGuardedCallbackDev
node_modules/react-dom/cjs/react-dom.development.js (4213:0)
invokeGuardedCallback
node_modules/react-dom/cjs/react-dom.development.js (4277:0)
beginWork$1
node_modules/react-dom/cjs/react-dom.development.js (27451:0)
performUnitOfWork
node_modules/react-dom/cjs/react-dom.development.js (26557:0)
workLoopSync
node_modules/react-dom/cjs/react-dom.development.js (26466:0)
renderRootSync
node_modules/react-dom/cjs/react-dom.development.js (26434:0)
performConcurrentWorkOnRoot
node_modules/react-dom/cjs/react-dom.development.js (25738:0)
workLoop
node_modules/scheduler/cjs/scheduler.development.js (266:0)
flushWork
node_modules/scheduler/cjs/scheduler.development.js (239:0)
MessagePort.performWorkUntilDeadline
node_modules/scheduler/cjs/scheduler.development.js (533:0)

**Comment by Roland-Hufnagel** [comment]
Hi Selco. A first step must be: 
In the FormCard component move the ImageUpload Component down below the ``</form>``. You can not nest a form inside another form. Please adapt this first step and push the changes to your repo. I don't have your cloudinary credentials so I cannot test it furthermore. Please adapt your code and try it out. If the error still occurs we'll see what the next steps could be.

**Reply by SelcoD** [reply]
Hi Roland, i put the ImageUpload under the Form, in the Detail components i commented it because if i want to render the Image, i get an Error to.
The error on the Form page dosent exist anymore but if i click submit, nothing happens

**Comment by Roland-Hufnagel** [comment]
Can you send me your credentials (.env.local) via slack?

**Comment by Roland-Hufnagel** [comment]
Hi Selco.

your last bug is: 

 In ImageUpload (line 32) you should import and use the next-Image component ``<Image>``with an uppercase I

**Reply by SelcoD** [reply]
<div>
      <form onSubmit={handleSubmit}>
        <label htmlFor="image-upload">Picture</label>
        <input type="file" name="image" id="image-upload" /> <-------THIS ONE ?
        <button type="submit">Submit</button>
      </form>
      <div>
        {tripImages.map((tripImage) => (
          <Image
            key={tripImage.id}
            src={tripImage.src}
            width={tripImage.width}
            height={tripImage.height}
            alt="Trip image"
            
            Like this ?
             return (
    <div>
      <form onSubmit={handleSubmit}>
        <label htmlFor="Image-upload">Picture</label>
        <input type="file" name="Image" id="Image-upload" /> <-----LIKE THIS ?
        <button type="submit">Submit</button>
      </form>
      <div>
        {tripImages.map((tripImage) => (
          <Image
            key={tripImage.id}
            src={tripImage.src}
            width={tripImage.width}
            height={tripImage.height}
            alt="Trip image"

**Comment by Roland-Hufnagel** [comment]
![Screenshot 2023-07-24 at 18 16 33](https://github.com/neuefische/questions/assets/111574974/b93b78dd-04be-49f9-967a-13142dfa7630)

**Reply by SelcoD** [reply]
changed it but now vercel has an deployment error

[PR](https://github.com/SelcoD/CampingFriends/pull/15)

**Comment by Roland-Hufnagel** [comment]
Maybe vercel needs the environment-variables?

**Comment by actyvystom** [comment, marked as answer]
Hi @SelcoD,
I found a few issues with your setup:
1. there is a typo in your upload api route handler function, needs to be called `handler` (instead of `hanlder`)
2. upload is working as expected, but there are some issues with your `ImageUpload` component:

you need to import the next image component:

```js
import Image from "next/image";
```

You are wrongly passing Image as a prop into your component:

```js
export default function ImageUpload(Image) {
```
Besides a wrong prop notation, it is not necessary to pass an image into that component as you are using the uploaded image for your display. So please change to

```js
export default function ImageUpload() {
```  
Then using the next/image component, you can add it inside your mapping over your `tripImages`: 
```js
{tripImages.map((tripImage) => (
          <Image
            key={tripImage.id}
            src={tripImage.src}
            width={tripImage.width}
            height={tripImage.height}
            alt="Trip image"
          />
        ))}
```
I strongly suggest to use a styled component for your next/image to ensure a fitting display for your App, e.g.
```js
const StyledImage = styled(Image)`
  border-radius: 0.5rem;
  width: 50vw;
  object-fit: contain;
  height: auto;
  border-color: aliceblue;
`;
```  
Using it in your mapping then:
```js
{tripImages.map((tripImage) => (
          <StyledImage
            key={tripImage.id}
            src={tripImage.src}
            width={tripImage.width}
            height={tripImage.height}
            alt="Trip image"
          />
        ))}
```

Hope that helps!
Thomas

**Reply by SelcoD** [reply]
Hello, thanks for your help but it still don't work. If i click on submit, nothing happens. Any other ideas ?
Thank you

**Reply by SelcoD** [reply]
<img width="1291" alt="Bildschirmfoto 2023-07-25 um 14 53 46" src="https://github.com/neuefische/questions/assets/132760864/29036e14-db6a-4e1a-93ce-dd1b59b718f1">

Maybe som issues with my vercel (localhost) ?

**Reply by actyvystom** [reply]
Looks like you set up your api key as a string... Don't put the values in environment variables in double quotes, just plain as they are, e.g.
```
CLOUDINARY_API_KEY=1234567890
```

**Reply by SelcoD** [reply]
<img width="1291" alt="Bildschirmfoto 2023-07-25 um 14 53 46" src="https://github.com/neuefische/questions/assets/132760864/00f40b56-a276-450f-8981-45f24d827127">
